src/cls/IPM/ResourceProcessor/WebApplication.cls

```diff
@@ -0,0 +1,263 @@
+    }
+    Return matchRoles
+}
+
+Method ReplaceNameSpace(expression As %String, namespace As %String) As %String
+{
+    Set templates = $Listbuild("{$ns}", "${ns}")
+    Set ptr = 0
+    While $LISTNEXT(templates, ptr, template) {
+        If expression[template {
+            Set expression = $REPLACE(expression, template, namespace)
+        }
+    }
+    Return expression
+}
+
+/// Called as phase <var>pPhase</var> is executed for the resource. If <var>pResourceHandled</var> is set to true,
+/// then the default behavior for that resource will be bypassed in the current phase.
+Method OnPhase(pPhase As %String, ByRef pParams, Output pResourceHandled As %Boolean = 0) As %Status
+{
+    Set sc = $$$OK
+    Try {
+        If (pPhase = "Validate") {
+            If (..Name = "") {
+                Return $$$ERROR($$$GeneralError,"Name is required for web applications")
+            }
+            Set keyList = ""
+            Set key = ""
+            While 1 {
+                Set value = ..ResourceReference.Attributes.GetNext(.key)
+                If (key = "") {
+                    Quit
+                }
+                Set keyList = keyList _ $ListBuild(key)
+            }
+            Set badKeyList = ..DoPropertiesExist(keyList)
+            If ($ListLength(badKeyList) > 0) {
+                Set error = "The following invalid attributes were found for resource "_..ResourceReference.Name_": "_$ListToString(badKeyList)
+                $$$ThrowStatus($$$ERROR($$$GeneralError,error))
+            }
+        }
+    } Catch (ex) {
+        Set sc = ex.AsStatus()
+    }
+    Return sc
+}
+
+Method OnBeforePhase(pPhase As %String, ByRef pParams) As %Status
+{
+    /*
+     * NOTE: Web app creation is done in OnBeforePhase so it happens before other
+     * resource processors such as FileCopy run.
+     */
+    Set sc = $$$OK
+    Try {
+        $$$ThrowOnError(##super(pPhase,.pParams))
+        
+        Set tVerbose = $Get(pParams("Verbose"))
+        #; Set tIsApplication = ..ResourceReference.Module.Lifecycle.%IsA(##class(%ZHSLIB.PackageManager.Developer.Lifecycle.Application).%ClassName(1))
+        Set tIsApplication = ..ResourceReference.Module.Lifecycle.%IsA(##class(%IPM.Lifecycle.DeployedModule).%ClassName(1))
+        
+        // TODO: Make mirror-safe?
+        // Would require moving mirror-safe APIs to the package manager - at which point, why not just have the package manager manage the whole federation?
+        If ((pPhase = "Configure") || ((pPhase = "Activate") && tIsApplication)) {
+            // Create web application
+            Do ..CreateOrUpdateWebApp(tVerbose)
+        } ElseIf ((pPhase = "Unconfigure") || ((pPhase = "Clean") && tIsApplication)) {
+            // Remove web application
+            Do ..DeleteWebApp(tVerbose)
+        }
+    } Catch e {
+        Set sc = e.AsStatus()
+    }
+    Quit sc
+}
+
+Method CreateOrUpdateWebApp(pVerbose As %Boolean = 0) [ Internal ]
+{
+    // TODO: Make mirror-safe.
```

Comment:
@isc-kiyer do you think that making this mirror-safe needs to happen as part of initial implementation? (Same for other places with the comment.)

@isc-shuliu for context, this would mean using https://docs.intersystems.com/irisforhealthlatest/csp/documatic/%25CSP.Documatic.cls?LIBRARY=HSLIB&CLASSNAME=HS.HC.SystemConfig.API if available. (This was renamed in I4H 2023.1; previously it was https://docs.intersystems.com/irisforhealth20221/csp/documatic/%25CSP.Documatic.cls?LIBRARY=HSLIB&CLASSNAME=HS.Util.SystemAPI. Not sure how far back we'd need to go.)

Reply:
@isc-tleavitt @isc-shuliu  Nope do not bother making it mirror safe. The IRIS security database will eventually support mirroring (hopefully by next year)

Reply:
Oooooh beautiful! That's news to me but great to hear.